File: frontend/src/components/Validity.js
// frontend/src/components/Validity.js
import React, { useEffect, useState } from 'react';
import axios from 'axios';
import './Validity.css';
import { getAuthHeaders } from '../utils/auth';

function Validity() {
  const [validityChecks, setValidityChecks] = useState([]);
  const [selectedCheck, setSelectedCheck] = useState(null);

  useEffect(() => {
    fetchValidityChecks();
  }, [userId]);

  const fetchValidityChecks = async () => {
    try {
      const headers = getAuthHeaders();
      const response = await axios.get(`${process.env.REACT_APP_API_URL}/api/validity-checks`, {
        headers: headers,
        params: {
          user_id: userId
        }
      });
      setValidityChecks(response.data.validityChecks);
    } catch (error) {
      console.error('Error fetching validity checks:', error);
      if (error.response) {
        setValidityChecks([{ error: error.response.data.error }]);
      } else {
        setValidityChecks([{ error: 'An unexpected error occurred' }]);
      }
    }
  };

  const handleCheckClick = (check) => {
    setSelectedCheck(check);
  };

  const handleDeleteCheck = async (checkId) => {
    try {
      const headers = getAuthHeaders();
      await axios.delete(`${process.env.REACT_APP_API_URL}/api/validity-checks/${checkId}`, {
        headers: headers,
        params: {
          user_id: userId
        }
      });
      fetchValidityChecks(); // Refresh the list after deletion
      if (selectedCheck && selectedCheck.id === checkId) {
        setSelectedCheck(null);
      }
    } catch (error) {
      console.error('Error deleting validity check:', error);
    }
  };

  const formatCheck = (check) => {
    return {
      ...check,
      review: check.review || 'N/A',
      style_guide_adherence: check.style_guide_adherence ? 'Yes' : 'No',
      style_guide_feedback: check.style_guide_adherence_feedback || 'N/A',
      continuity: check.continuity ? 'Yes' : 'No',
      continuity_feedback: check.continuity_feedback || 'N/A',
      test_results: check.test_results || 'N/A'
    };
  };

  return (
    <div className="validity-container">
      <div className="validity-sidebar">
        <h2>Validity Checks</h2>
        <ul>
          {validityChecks.map((check) => (
            <li
              key={check.id}
              className={`validity-check ${selectedCheck && selectedCheck.id === check.id ? 'selected' : ''}`}
              onClick={() => handleCheckClick(check)}
            >
              {check.chapterTitle} : {check.isValid ? 'Valid' : 'Invalid'}
              <span className="remove-icon" onClick={() => handleDeleteCheck(check.id)}>❌</span>
            </li>
          ))}
        </ul>
      </div>
      <div className="validity-content">
        {selectedCheck && (
          <div className="validity-details">
            <h3>{selectedCheck.chapterTitle}</h3>
            <p>Validity: {selectedCheck.isValid ? 'Valid' : 'Invalid'}</p>
            <div className="validity-details-section">
              <p><strong>Chapter ID:</strong> {selectedCheck.chapterId}</p>
              <p><strong>Chapter Title:</strong> {selectedCheck.chapterTitle}</p>
              <p><strong>Validity:</strong> {selectedCheck.isValid ? 'Valid' : 'Invalid'}</p>
              <p><strong>Feedback:</strong> {selectedCheck.feedback}</p>
              <p><strong>Review:</strong> {formatCheck(selectedCheck).review}</p>
              <p><strong>Style Guide Adherence:</strong> {formatCheck(selectedCheck).style_guide_adherence}</p>
              <p><strong>Style Guide Feedback:</strong> {formatCheck(selectedCheck).style_guide_feedback}</p>
              <p><strong>Continuity:</strong> {formatCheck(selectedCheck).continuity}</p>
              <p><strong>Continuity Feedback:</strong> {formatCheck(selectedCheck).continuity_feedback}</p>
              <p><strong>Test Results:</strong> {formatCheck(selectedCheck).test_results}</p>
            </div>
          </div>
        )}
      </div>
    </div>
  );
}

export default Validity;

File: flutterfrontend/lib/components/validity.dart
import 'package:flutter/material.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import '../utils/auth.dart';
import '../utils/constants.dart';
import 'package:fluttertoast/fluttertoast.dart';

class Validity extends StatefulWidget {
  const Validity({Key? key}) : super(key: key);

  @override
  State<Validity> createState() => _ValidityState();
}

class _ValidityState extends State<Validity> {
  List<dynamic> _validityChecks = [];
  dynamic _selectedCheck;

  @override
  void initState() {
    super.initState();
    _fetchValidityChecks();
  }

  Future<void> _fetchValidityChecks() async {
    try {
      final response = await http.get(
        Uri.parse('$apiUrl/validity-checks'),
        headers: await getAuthHeaders(),
      );
      if (response.statusCode == 200) {
        setState(() {
          _validityChecks = json.decode(response.body)['validityChecks'];
        });
      } else {
        Fluttertoast.showToast(msg: 'Error fetching validity checks');
      }
    } catch (error) {
      print('Error fetching validity checks: $error');
      Fluttertoast.showToast(msg: 'Error fetching validity checks');
    }
  }

  void _handleCheckClick(dynamic check) {
    setState(() {
      _selectedCheck = check;
    });
  }

  Future<void> _handleDeleteCheck(String checkId) async {
    try {
      final response = await http.delete(
        Uri.parse('$apiUrl/validity-checks/$checkId'),
        headers: await getAuthHeaders(),
      );
      final jsonResponse = json.decode(response.body);
      if (response.statusCode == 204 && !jsonResponse.containsKey('error')) {
        setState(() {
          _validityChecks.removeWhere((check) => check['id'] == checkId);
          if (_selectedCheck != null && _selectedCheck['id'] == checkId) {
            _selectedCheck = null;
          }
        });
        Fluttertoast.showToast(msg: 'Validity check deleted successfully');
      } else {
        final errorMessage = jsonResponse['error'] ?? 'Error deleting validity check';
        Fluttertoast.showToast(msg: errorMessage);
      }
    } catch (error) {
      print('Error deleting validity check: $error');
      Fluttertoast.showToast(msg: 'Error deleting validity check');
    }
  }

  String _formatBool(bool? value) => value == null
      ? 'N/A'
      : value
          ? 'Yes'
          : 'No';
  String _formatString(String? value) => value ?? 'N/A';

  @override
  Widget build(BuildContext context) {
    return ListView.builder(
      itemCount: _validityChecks.length,
      padding: const EdgeInsets.all(16),
      itemBuilder: (context, index) {
        final check = _validityChecks[index];
        return Card(
          elevation: 3,
          margin: const EdgeInsets.symmetric(vertical: 8),
          child: ListTile(
            title: Text(
              '${check['chapterTitle']}: ${check['isValid'] ? 'Valid' : 'Invalid'}',
              style: const TextStyle(
                fontWeight: FontWeight.bold,
                fontSize: 18,
              ),
            ),
            onTap: () => showDialog(
              context: context,
              builder: (BuildContext context) {
                return AlertDialog(
                  title: Text(check['chapterTitle']),
                  content: SingleChildScrollView(
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Text('Validity: ${check['isValid'] ? 'Valid' : 'Invalid'}'),
                        Text('Chapter ID: ${check['chapterId']}'),
                        Text('Feedback: ${_formatString(check['feedback'])}'),
                        Text('Review: ${_formatString(check['review'])}'),
                        Text('Style Guide Adherence: ${_formatBool(check['style_guide_adherence'])}'),
                        Text('Style Guide Feedback: ${_formatString(check['style_guide_feedback'])}'),
                        Text('Continuity: ${_formatBool(check['continuity'])}'),
                        Text('Continuity Feedback: ${_formatString(check['continuity_feedback'])}'),
                        Text('Test Results: ${_formatString(check['test_results'])}'),
                      ],
                    ),
                  ),
                  actions: [
                    TextButton(
                      onPressed: () => Navigator.pop(context),
                      child: const Text('Close'),
                    ),
                  ],
                );
              },
            ),
            trailing: IconButton(
              icon: const Icon(Icons.delete),
              onPressed: () => _handleDeleteCheck(check['id']),
            ),
          ),
        );
      },
    );
  }
}
